core/src/core/block.rs

```diff
@@ -289,48 +296,52 @@ impl Writeable for BlockHeader {
 	}
 }
 
+fn read_block_header(reader: &mut dyn Reader) -> Result<BlockHeader, ser::Error> {
+	let version = HeaderVersion::read(reader)?;
+	let (height, timestamp) = ser_multiread!(reader, read_u64, read_i64);
+	let prev_hash = Hash::read(reader)?;
+	let prev_root = Hash::read(reader)?;
+	let output_root = Hash::read(reader)?;
+	let range_proof_root = Hash::read(reader)?;
+	let kernel_root = Hash::read(reader)?;
+	let total_kernel_offset = BlindingFactor::read(reader)?;
+	let (output_mmr_size, kernel_mmr_size) = ser_multiread!(reader, read_u64, read_u64);
+	let pow = ProofOfWork::read(reader)?;
+
+	if timestamp > MAX_DATE.and_hms(0, 0, 0).timestamp()
+		|| timestamp < MIN_DATE.and_hms(0, 0, 0).timestamp()
+	{
+		return Err(ser::Error::CorruptedData);
+	}
+
+	// Check the block version before proceeding any further.
+	// We want to do this here because blocks can be pretty large
+	// and we want to halt processing as early as possible.
+	// If we receive an invalid block version then the peer is not on our hard-fork.
+	if !consensus::valid_header_version(height, version) {
```

Comment:
Is the `valid_header_version` check worth moving into `UntrustedBlockHeader`?
It takes a read lock on the global CHAIN_TYPE internally and we presumably don't need to do this check every time we read a header from the db.

Reply:
Sounds reasonable, will move it there

Reply:
done